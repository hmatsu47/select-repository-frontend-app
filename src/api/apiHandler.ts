import { fetchWithTimeout } from "./fetchWithTimeout";

// ベース URI
export const baseUri = "http://10.11.17.43/api";

// APIからデータ取得（GET）
export const getApiData = async (url: string) => {
  return await (await fetchWithTimeout(url)).json();
};
// APIからデータ取得（POST）
export const postApiData = async (url: string, object: Object) => {
  return await (await fetchWithTimeout(url, object)).json();
};
