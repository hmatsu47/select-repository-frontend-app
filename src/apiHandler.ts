import { fetchWithTimeout } from "./fetchWithTimeout";

// APIからデータ取得（GET）
export const getApiData = async (url: string) => {
  return await (await fetchWithTimeout(url)).json();
};
// APIからデータ取得（POST）
export const postApiData = async (url: string, object: Object) => {
  return await (await fetchWithTimeout(url, object)).json();
};
