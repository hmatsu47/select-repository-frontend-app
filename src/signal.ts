import { createSignal } from "solid-js";
import { AlertColor } from "@suid/material/Alert";

import { ImageItem, RepositoryItem, ServiceItem } from "./type";

export const [service, setService] = createSignal<string | undefined>();
export const [repository, setRepository] = createSignal<string | undefined>();
export const [services, setServices] = createSignal<
  ServiceItem[] | undefined
>();
export const [repositories, setRepositories] = createSignal<
  RepositoryItem[] | undefined
>();
export const [images, setImages] = createSignal<ImageItem[] | undefined>();
export const [imageUri, setImageUri] = createSignal<string | undefined>();
export const [releaseAt, setReleaseAt] = createSignal<Date | undefined>();
export const [lastImageUri, setLastImageUri] = createSignal<
  string | undefined
>();
export const [lastReleasedAt, setLastReleasedAt] = createSignal<
  Date | undefined
>();
export const [isReleased, setIsReleased] = createSignal<boolean | undefined>();
export const [message, setMessage] = createSignal<string | undefined>();
export const [messageSeverity, setMessageSeverity] = createSignal<
  AlertColor | undefined
>();
