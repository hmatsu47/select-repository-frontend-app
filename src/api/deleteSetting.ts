import { ErrorResponse } from "../type";
import { baseUri, deleteApiData } from "./apiHandler";
import {
  imageUri,
  isReleased,
  isReleaseSelected,
  releaseAt,
  service,
  setMessage,
  setMessageSeverity,
  setImageUri,
  setIsReleaseSelected,
  setReleaseAt,
} from "../signal";

export const deleteSetting = async () => {
  if (
    !isReleased() === undefined ||
    !isReleaseSelected() ||
    !imageUri() ||
    !releaseAt()
  ) {
    return;
  }
  const load = async (): Promise<void> => {
    const data: null | ErrorResponse = await deleteApiData(
      `${baseUri}/setting/${service()}`
    );
    if (
      typeof data === "object" &&
      data !== null &&
      typeof (data as ErrorResponse).message === "string"
    ) {
      // 戻り値がエラーメッセージの場合
      setMessage((data as ErrorResponse).message);
      setMessageSeverity("error");
      return;
    }
    setImageUri(undefined);
    setReleaseAt(undefined);
    setIsReleaseSelected(false);
    setMessage("リリースを取り消しました。");
    setMessageSeverity("success");
  };
  void load();
};
