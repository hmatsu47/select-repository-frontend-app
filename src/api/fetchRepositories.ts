import { ErrorResponse, RepositoryItem } from "../type";
import { baseUri, getApiData } from "./apiHandler";
import {
  service,
  setImages,
  setMessage,
  setMessageSeverity,
  setRepositories,
} from "../signal";

export const fetchRepositories = async () => {
  if (!service()) {
    return;
  }
  const load = async (): Promise<void> => {
    const data: RepositoryItem[] | ErrorResponse = await getApiData(
      `${baseUri}/repositories/${service()}`
    );
    if (
      typeof data === "object" &&
      data !== null &&
      typeof (data as ErrorResponse).message === "string"
    ) {
      // 戻り値がエラーメッセージの場合
      setRepositories(undefined);
      setImages(undefined);
      setMessage((data as ErrorResponse).message);
      setMessageSeverity("error");
      return;
    }
    setRepositories(data as RepositoryItem[]);
    setImages(undefined);
    setMessage(undefined);
  };
  void load();
};
