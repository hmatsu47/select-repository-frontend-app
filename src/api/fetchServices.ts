import { ErrorResponse, ServiceItem } from "../type";
import { baseUri, getApiData } from "./apiHandler";
import {
  setErrorMessage,
  setImages,
  setRepositories,
  setServices,
} from "../signal";

export const fetchServices = async () => {
  const load = async (): Promise<void> => {
    const data: ServiceItem[] | ErrorResponse = await getApiData(
      `${baseUri}/services`
    );
    if (
      typeof data === "object" &&
      data !== null &&
      typeof (data as ErrorResponse).message === "string"
    ) {
      // 戻り値がエラーメッセージの場合
      setServices(undefined);
      setRepositories(undefined);
      setImages(undefined);
      setErrorMessage((data as ErrorResponse).message);
      return;
    }
    setServices(data as ServiceItem[]);
    setRepositories(undefined);
    setImages(undefined);
    setErrorMessage(undefined);
  };
  void load();
};
