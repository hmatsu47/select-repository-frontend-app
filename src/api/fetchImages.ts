import { ErrorResponse, ImageItem } from "../type";
import { baseUri, getApiData } from "./apiHandler";
import { repository, service, setErrorMessage, setImages } from "../signal";

export const fetchImages = async () => {
  if (!service() || !repository()) {
    return;
  }
  const load = async (): Promise<void> => {
    const data: ImageItem[] | ErrorResponse = await getApiData(
      `${baseUri}/images/${service()}/${repository()}`
    );
    if (
      typeof data === "object" &&
      data !== null &&
      typeof (data as ErrorResponse).message === "string"
    ) {
      // 戻り値がエラーメッセージの場合
      setImages(undefined);
      setErrorMessage((data as ErrorResponse).message);
      return;
    }
    setImages(data as ImageItem[]);
    setErrorMessage(undefined);
  };
  void load();
};
