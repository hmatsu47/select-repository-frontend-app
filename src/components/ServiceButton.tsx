import Button from "@suid/material/Button";
import { service, setService } from "../signal";
import { fetchRepositories } from "../api/fetchRepositories";
import { fetchSetting } from "../api/fetchSetting";
import { ServiceItem } from "../type";

type Props = {
  serviceItem: ServiceItem;
};

export const ServiceButton = (props: Props) => {
  return (
    <Button
      variant={service() === props.serviceItem.name ? "contained" : "outlined"}
      size="small"
      color="inherit"
      onClick={async (e) => {
        setService(props.serviceItem.name);
        await fetchSetting();
        await fetchRepositories();
      }}
      sx={{ textTransform: "none" }}
    >
      {props.serviceItem.name}
    </Button>
  );
};
