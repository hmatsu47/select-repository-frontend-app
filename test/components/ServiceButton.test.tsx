// @vitest-environment jsdom

import { describe, expect, test, vi } from "vitest";
import "vi-fetch/setup";
import { mockFetch, mockGet } from "vi-fetch";
import { fireEvent, render } from "solid-testing-library";
import { formatSnapshot } from "../common/formatSnapshot";
import { RepositoryItem, ServiceItem, Setting } from "../../src/type";
import { baseUri } from "../../src/api/apiHandler";
import { ServiceButton } from "../../src/components/ServiceButton";
import { service } from "../../src/signal";

describe("<ServiceButton />", () => {
  beforeEach(() => {
    mockFetch.clearAll();
  });
  afterEach(() => {
    vi.clearAllMocks();
  });
  test("サービスボタン", async () => {
    localStorage.removeItem("selectedService");
    const mockSetting = mockGet(`${baseUri}/setting/service1`).willResolve({
      is_released: false,
    } as Setting);
    const mockRepositories = mockGet(
      `${baseUri}/repositories/service1`
    ).willResolve([
      {
        name: "repository1",
        uri: "000000000000.dkr.ecr.ap-northeast-1.amazonaws.com/repository1",
      } as RepositoryItem,
    ]);
    const { container, findByText, unmount } = render(() => (
      <ServiceButton serviceItem={{ name: "service1" } as ServiceItem} />
    ));
    // css の名前が動的に変わるので固定値に置換
    const htmlBefore = formatSnapshot(container.innerHTML);
    expect(htmlBefore).toMatchSnapshot();
    // ボタンクリック
    const button = (await findByText("service1")) as HTMLInputElement;
    expect(button).toHaveTextContent("service1");
    fireEvent.click(button);
    // ボタンクリック後の照合
    const htmlAfter = formatSnapshot(container.innerHTML);
    expect(htmlAfter).toMatchSnapshot();
    expect(mockSetting).toHaveFetched();
    expect(service()).toBe("service1");
    // 現状では setting の呼び出しまでしか捕捉できないようなのでコメントアウト
    // expect(mockRepositories).toHaveFetched();
    unmount();
    localStorage.removeItem("selectedService");
  });
});
