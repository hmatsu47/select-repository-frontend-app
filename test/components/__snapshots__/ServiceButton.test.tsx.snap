// Vitest Snapshot v1

exports[`<ServiceButton /> > サービスボタン 1`] = `"<button type=\\"button\\" class=\\"MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedInherit MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorInherit MuiButton-root MuiButtonBase-root css-xxxxxxxx\\" title=\\"service1 を選択\\" tabindex=\\"0\\">service1<span class=\\"MuiTouchRipple-root css-xxxxxxxx\\"></span></button>"`;

exports[`<ServiceButton /> > サービスボタン 2`] = `"<button type=\\"button\\" class=\\"MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-root MuiButtonBase-root css-xxxxxxxx_1\\" title=\\"service1 を選択\\" tabindex=\\"0\\">service1<span class=\\"MuiTouchRipple-root css-xxxxxxxx\\"></span></button>"`;
